datafusion/core/src/physical_planner.rs

```diff
@@ -496,776 +550,966 @@ impl DefaultPhysicalPlanner {
-                    if let Some(provider) = schema.table(name).await? {
-                        let input_exec = self.create_initial_plan(input, session_state).await?;
-                        provider.insert_into(session_state, input_exec, false).await
-                    } else {
-                        return exec_err!(
-                            "Table '{table_name}' does not exist"
-                        );
-                    }
-                }
-                LogicalPlan::Dml(DmlStatement {
-                    table_name,
-                    op: WriteOp::InsertOverwrite,
-                    input,
-                    ..
-                }) => {
-                    let name = table_name.table();
-                    let schema = session_state.schema_for_ref(table_name.clone())?;
-                    if let Some(provider) = schema.table(name).await? {
-                        let input_exec = self.create_initial_plan(input, session_state).await?;
-                        provider.insert_into(session_state, input_exec, true).await
-                    } else {
-                        return exec_err!(
-                            "Table '{table_name}' does not exist"
-                        );
-                    }
+                    // Indices refer to position in flat tree Vec, which means they are
+                    // guaranteed to be unique, hence unstable sort used.
+                    //
+                    // We reverse sort because of how we visited the node in the initial
+                    // DFS traversal (see above).
+                    children.sort_unstable_by_key(|(index, _)| std::cmp::Reverse(*index));
+                    let children =
+                        children.iter().map(|(_, plan)| plan.clone()).collect();
+                    let children = ChildrenContainer::Multiple(children);
+                    plan = self
+                        .map_logical_node_to_physical(node.node, session_state, children)
+                        .await?;
                 }
-                LogicalPlan::Values(Values {
-                    values,
-                    schema,
-                }) => {
-                    let exec_schema = schema.as_ref().to_owned().into();
-                    let exprs = values.iter()
-                        .map(|row| {
-                            row.iter().map(|expr| {
-                                self.create_physical_expr(
-                                    expr,
-                                    schema,
-                                    session_state,
-                                )
+            }
+            current_index = parent_index;
+        }
+        // Only one task should ever reach this point for a valid LogicalPlan tree.
+        Ok(Some(plan))
+    }
+
+    /// Given a single LogicalPlan node, map it to it's physical ExecutionPlan counterpart.
+    async fn map_logical_node_to_physical(
```

Comment:
As I understand it the main reason this structure reduces stack space is that this function requires a non trivial stack frame, but now instead of recursively calling itself (which results in many such frames on the stack) it calls itself iteratively (basically pushing the results on to a Vec)

Or put another way, `map_logical_node_to_physical` never calls `map_logical_node_to_physical`